software/openTun/openTun.c

```diff
@@ -0,0 +1,920 @@
+  nr.rssi    = inBuf[28];
+  nr.numRx   = inBuf[29];
+  nr.numTx   = inBuf[30];
+  nr.numTxACK= inBuf[31];
+  nr.numWraps= inBuf[32];
+  nr.asn[0]  = inBuf[33];
+  nr.asn[2]  = inBuf[34];
+  nr.asn[1]  = inBuf[35];
+  nr.asn[4]  = inBuf[36];
+  nr.asn[3]  = inBuf[37];
+  nr.joinPrio= inBuf[38];
+
+  if(nr.row < MAX_NEIGHBORS_ROW) {
+    stats.neighbor_rows[nr.row] = nr;
+    if(stats.neighbor_rows_max <= nr.row) {
+      stats.neighbor_rows_max = nr.row;
+    }
+  }
+};
+
+void parse_joinHistoryRow(unsigned char inBuf[], unsigned int inLen)
+{
+  struct joinedHistory *jh;
+  unsigned int previousRow;
+
+  /* guard against too small packets */
+  if(inLen < 5) return;
+
+#if 0
+  if(stats.joined_history_max == 0) {
+    printf("WAITING FOR DEBUGGER\n");
+    sleep(60);
+  }
+#endif
+
+  if(stats.joined_history_cur > MAX_JOINHIST_ROW) {
+    stats.joined_history_cur=0;
+  }
+
+  previousRow = stats.joined_history_cur-1;
+  if(stats.joined_history_cur == 0) {
+    previousRow = stats.joined_history_max-1;
+  }
+
+  /* only log a new join ASN if it is different than previous one */
+  if(stats.joined_history_max > 0 &&
+     memcmp(stats.joinedHistory_rows[previousRow].asn,
+            inBuf, 5)==0) {
+    return;
+  }
+
+  jh = &stats.joinedHistory_rows[stats.joined_history_cur];
+
+  jh->row    = stats.joined_count;
+  jh->asn[0] = inBuf[0];
+  jh->asn[1] = inBuf[1];
+  jh->asn[2] = inBuf[2];
+  jh->asn[3] = inBuf[3];
+  jh->asn[4] = inBuf[4];
+
+  stats.joined_history_cur++;
+  stats.joined_count++;
+  if(stats.joined_history_max < stats.joined_history_cur) {
+    stats.joined_history_max = stats.joined_history_cur;
+  }
+
+}
+
+void parse_status(unsigned char inBuf[], unsigned int inLen)
+{
+  if(inLen < 4) return;
+  unsigned int moteId = inBuf[1] + inBuf[2]<<8;
+  unsigned int statusElem = inBuf[3];
+
+  if(verbose > 2) fprintf(stderr, "moteId %04x status: %u --:", moteId, statusElem);
+  switch(statusElem) {
+  case 0:
+    dump_screen(&stats);
+    break;
+
+  case 1:
+    parse_idmanager(inBuf, inLen);
+    break;
+
+  case 2:
+    stats.myDAGrank = inBuf[4];
+    break;
+
+  case 3:
+    if(inLen >= 8) {
+      stats.index_read = inBuf[4] + inBuf[5]<<8;
+      stats.index_write= inBuf[6] + inBuf[7]<<8;
+    }
+    break;
+
+  case 4:
+    if(inLen >= 5 + 4) {
+      stats.asn[0]  = inBuf[4];
+      stats.asn[2]  = inBuf[5];
+      stats.asn[1]  = inBuf[6];
+      stats.asn[4]  = inBuf[7];
+      stats.asn[3]  = inBuf[8];
+    }
+    break;
+
+  case 5:
+    parse_macstats(inBuf, inLen);
+    break;
+
+  case 6:
+    if(verbose > 2) fprintf(stderr, "ScheduleRow\n");
+    parse_scheduleRow(inBuf, inLen);
+    break;
+
+  case 7:
+    if(inLen >= 6) {
+      stats.backoffExponent = inBuf[4];
+      stats.backoff =         inBuf[5];
+    }
+    break;
+
+  case 8:
+    if(verbose > 2) fprintf(stderr, "QueueRow\n");
+    parse_queueRow(inBuf, inLen);
+    break;
+
+  case 9:
+    if(verbose) fprintf(stderr, "NeighborsRow\n");
```

Comment:
I see that I didn't call it yet, so maybe it doesn't work yet.  I will validate that it works and send more code later.